src/pretix/plugins/paypal/views.py

```diff
@@ -76,40 +111,134 @@ def redirect_view(request, *args, **kwargs):
+                    pass
+                else:
+                    order = None
+        else:
+            order = None
+
+        # urlkwargs = {}
+        # if 'cart_namespace' in kwargs:
+        #     urlkwargs['cart_namespace'] = kwargs['cart_namespace']
+
+        prov = PaypalMethod(request.event)
+
+        if order:
+            lp = order.payments.last()
+            if lp and lp.state not in (OrderPayment.PAYMENT_STATE_CONFIRMED, OrderPayment.PAYMENT_STATE_REFUNDED):
+                fee = lp.fee.value - prov.calculate_fee(order.pending_sum - lp.fee.value)
+            else:
+                fee = prov.calculate_fee(order.pending_sum)
+
+            cart = {
+                'positions': order.positions,
+                'total': order.pending_sum,
+                'fee': fee,
+            }
+        else:
+            cart = {
+                'positions': get_cart(request),
+                'total': get_cart_total(request),
+                'fee': prov.calculate_fee(get_cart_total(request)),
+            }
+
+        paypal_order = prov._create_paypal_order(request, None, cart)
+        r = JsonResponse(paypal_order.dict())
+        r._csp_ignore = True
+        return r
+
+
+@method_decorator(xframe_options_exempt, 'dispatch')
+class PayView(PaypalOrderView, TemplateView):
+    template_name = ''
+
+    def get(self, request, *args, **kwargs):
+        if self.payment.state not in [OrderPayment.PAYMENT_STATE_CREATED, OrderPayment.PAYMENT_STATE_PENDING]:
+            return self._redirect_to_order()
+        else:
+            r = render(request, 'pretixplugins/paypal/pay.html', self.get_context_data())
+            return r
+
+    def post(self, request, *args, **kwargs):
+        self.payment.payment_provider.execute_payment(request, self.payment)
+        return self._redirect_to_order()
+
+    def get_context_data(self, **kwargs):
+        ctx = super().get_context_data(**kwargs)
+
+        ctx['order'] = self.order
+        ctx['oid'] = self.payment.info_data['id']
+        ctx['method'] = self.payment.payment_provider.method
+        return ctx
+
+
 @scopes_disabled()
-def oauth_return(request, *args, **kwargs):
-    if 'payment_paypal_oauth_event' not in request.session:
+def isu_return(request, *args, **kwargs):
+    getparams = ['merchantId', 'merchantIdInPayPal', 'permissionsGranted', 'accountStatus', 'consentStatus', 'productIntentID', 'isEmailConfirmed']
+    sessionparams = ['payment_paypal_isu_event', 'payment_paypal_isu_tracking_id']
+    if not any(k in request.GET for k in getparams) or not any(k in request.session for k in sessionparams):
         messages.error(request, _('An error occurred during connecting with PayPal, please try again.'))
         return redirect(reverse('control:index'))
 
-    event = get_object_or_404(Event, pk=request.session['payment_paypal_oauth_event'])
+    event = get_object_or_404(Event, pk=request.session['payment_paypal_isu_event'])
 
+    gs = GlobalSettingsObject()
```

Comment:
I know we also didn't have this in the past, but maybe it makes sense here to check if you're still logged in as a user that is allowed to manipulate the settings of this event? it would be too bad to have an attack where I can reconfigure your event to my paypal account and get all your payments just because I somehow got the session manipulated (should be impossible, but defense in depth, bla)